mautic/mautic
# Phpstan, Rector and JMS Serializer update

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [X]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

I wanted to update Phpstan and Rector to give us access to more rules. It was blocked by the JMS Serializer so I had to update it from 3.8 to 4.0. I believe the [BC changes in the JMS Serializer](https://github.com/schmittjoh/JMSSerializerBundle/blob/master/UPGRADING.md#upgrading-from-3x-to-40) do not apply to us as we do not use Twig nor the services so I do not take it as a BC change.

The Phpstan Baseline file had to be regenerated as the new version use slightly different error messages and so the regexes do not match those in the original file. There seem to be a few new issues found and some old removed as we've resolved many during the development.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. The only thing to test is API as that is using the JMS Serializer. But it is already tested many times in the Github Actions so green CI checks should be enough.




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11131"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
